cloudfront/wellcomelibrary.org/edge-lambda/src/wellcomeLibraryRedirect.ts

```diff
@@ -1,21 +1,62 @@
-import { CloudFrontRequestHandler } from 'aws-lambda';
-import { CloudFrontRequest } from 'aws-lambda/common/cloudfront';
-import {redirect} from "./redirect";
+import { CloudFrontRequestEvent, Context } from 'aws-lambda';
+import {
+  CloudFrontRequest,
+  CloudFrontResultResponse,
+} from 'aws-lambda/common/cloudfront';
+import { getBnumberFromPath } from './paths';
+import { getWork } from './bnumberToWork';
+import { createRedirect } from './createRedirect';
+import { redirectToRoot } from './redirectToRoot';
 
-export const requestHandler: CloudFrontRequestHandler = (event, context, callback) => {
-  const request: CloudFrontRequest = event.Records[0].cf.request;
+async function rewriteRequestUri(
+  uri: string
+): Promise<undefined | CloudFrontResultResponse> {
+  const itemPathRegExp: RegExp = /^\/item\/.*/;
+  const wellcomeCollectionHost = 'https://wellcomecollection.org';
+  const notFoundRedirect = createRedirect(
+    `${wellcomeCollectionHost}/works/not-found`
+  );
+
+  if (uri.match(itemPathRegExp)) {
+    // Try and find b-number in item path
+    const bNumberResult = getBnumberFromPath(uri);
+
+    if (bNumberResult instanceof Error) {
+      return notFoundRedirect;
+    }
 
-  // Redirect www. -> to root
-  if(request.headers.host && request.headers.host.length == 1) {
-    const requestHost =  request.headers.host[0].value
+    // Find corresponding work id
+    const bNumber = bNumberResult;
+    const work = await getWork(bNumber);
 
-    if (requestHost.startsWith('www.')) {
-      const rootRequestHost = requestHost.replace('www.','');
-      return Promise.resolve(redirect(`https://${rootRequestHost}${request.uri}`));
+    if (work instanceof Error) {
+      return notFoundRedirect;
```

Comment:
Will there be a way of distinguishing which of these 2 error cases happened?

Reply:
We may add metrics later, I should at least log these errors I think - will fix.